blog/2019-02-07-KIEServerOptaPlannerTaskAssignment.adoc

```diff
@@ -0,0 +1,215 @@
+----
+
+=== Persistence
+
+Since you will be sending and receiving data to/from KIE Server through REST API, you need to tell it how to marshall/unmarshall
+this data. Read how OptaPlanner
+https://docs.optaplanner.org/7.17.0.Final/optaplanner-docs/html_single/index.html#integrationWithXStream[marshals a score] using Xstream.
+
+OptaPlanner stores all planning entities as objects and references to these objects, which might result in a lot of redundant data
+received from the server. In our example, we are using
+https://github.com/FasterXML/jackson-annotations/wiki/Jackson-Annotations#object-references-identity[@JsonIdentityInfo]
+to avoid such redundancies.
+
+=== Dependencies
+
+The minimum required dependencies to build the OptaPlanner service are: optaplanner-core, optaplanner-persistence-xstream.
+
+TIP: Because you will be using several OptaPlanner modules, it’s recommended to import the optaplanner-bom
+in Maven’s `dependencyManagement` so that OptaPlanner version is specified only once.
+
+
+== Build and Deployment
+
+Now that you have the KJAR ready, you can deploy it to the KIE Server and perform all planning operations.
+You can send HTTP requests to the server using REST API.
+
+=== Deploymnet
+
+`POST /config`
+
+Through the above endpoint, you can execute various commands on the KIE Server.
+
+For example, to create a container running your OptaPlanner service, the request's body should be:
+[source,json,options="nowrap"]
+----
+{
+  commands: [
+    {
+      'create-container': {
+        'container': {
+          'container-id': CONTAINER_ID,
+          'release-id': {
+            'group-id': GROUP_ID,
+            'artifact-id': ARTIFACT_ID,
+            'version': VERSION,
+          },
+        },
+      },
+    },
+  ],
+}
+----
+Notice that the GAV in the `release-id` object are your KJAR's GAV.
+
+=== Solver registration
+
+`PUT /containers/{CONTAINER_ID}/solvers/{SOLVER_ID}`
+
+With body:
+[source,json,options="nowrap"]
+----
+{
+  'solver-config-file': 'PATH_TO_SOLVER_CONFIG_FILE.xml'
+}
+----
+This will build a new solver from the xml resource included in the KJAR.
+
+=== Submit a problem
+
+`POST /containers/{CONTAINER_ID}/solvers/{SOLVER_ID}/state/solving`
+
+Once a solver is built it will be waiting for a problem to start solving. The body of this request contains the object
+annotated as `@PlanningSolution`, in the Task Assignment example it will be `TaskAssigningSolution`.
+
+=== Query bestSolution
+
+`GET /containers/{CONTAINER_ID}/solvers/{SOLVER_ID}/bestsolution`
+
+This response body will contain `best-solution` object in addition to extra information about the solver status and score.
+
+=== Submit a ProblemFactChange
+
+`POST /containers/{CONTAINER_ID}/solvers/{SOLVER_ID}/problemfactchanges`
+
+Submits a ProblemFactChange to update the problem the solver is solving. For example if you want to delete
+a task the body should be:
+//TODO: Update the example to JSON once figure out the required format
```

Comment:
This might help you https://issues.jboss.org/browse/BXMSDOC-3370?focusedCommentId=13691712&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-13691712.

Reply:
Update: I don't know if your JSON format problem is related to `java.time.LocalDateTime`. If it's not, disregard my comment.

Reply:
The problem is how to tell the kjar which ProblemFactChange class is thiss, with xml this is achieved by using class attribute

Reply:
I see. This might be helpful: https://www.baeldung.com/jackson-inheritance.